In [16]:
pip install numpy

Note: you may need to restart the kernel to use updated packages.


In [17]:
import numpy as np

In [19]:
import pandas as pd

In [20]:
df = pd.read_csv('NSMES1988.csv')

In [21]:
df.info(memory_usage='deep')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4406 entries, 0 to 4405
Data columns (total 20 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Unnamed: 0  4406 non-null   int64  
 1   visits      4406 non-null   int64  
 2   nvisits     4406 non-null   int64  
 3   ovisits     4406 non-null   int64  
 4   novisits    4406 non-null   int64  
 5   emergency   4406 non-null   int64  
 6   hospital    4406 non-null   int64  
 7   health      4406 non-null   object 
 8   chronic     4406 non-null   int64  
 9   adl         4406 non-null   object 
 10  region      4406 non-null   object 
 11  age         4406 non-null   float64
 12  afam        4406 non-null   object 
 13  gender      4406 non-null   object 
 14  married     4406 non-null   object 
 15  school      4406 non-null   int64  
 16  income      4406 non-null   float64
 17  employed    4406 non-null   object 
 18  insurance   4406 non-null   object 
 19  medicaid    4406 non-null  

# 2.4 MB for Data Size
- 4406 rows and 20 columns
- data types include int64, objects, float64

In [22]:
df.shape[0]
print('Rows are', df.shape[0])
df.shape[1]
print('Columns are', df.shape[1])
df.dtypes

Rows are 4406
Columns are 20


Unnamed: 0      int64
visits          int64
nvisits         int64
ovisits         int64
novisits        int64
emergency       int64
hospital        int64
health         object
chronic         int64
adl            object
region         object
age           float64
afam           object
gender         object
married        object
school          int64
income        float64
employed       object
insurance      object
medicaid       object
dtype: object

In [23]:
df.isna().sum()

Unnamed: 0    0
visits        0
nvisits       0
ovisits       0
novisits      0
emergency     0
hospital      0
health        0
chronic       0
adl           0
region        0
age           0
afam          0
gender        0
married       0
school        0
income        0
employed      0
insurance     0
medicaid      0
dtype: int64

In [24]:
print('Minimum age of the data set is', df['age'].min()*10 , 'and the max age is' , df['age'].max()*10)

Minimum age of the data set is 66.0 and the max age is 109.0


- Minimum age of the data set is 66.0 and the max age is 109.0
- Minimum income of the data set is -$10,125 and the max income is $548,351
- notice of negative income for the minimum value


In [25]:
df['income'].min()*10**4

-10125.0

# Neg

In [26]:
df['income'].max()*10**4

548351.0

In [27]:
print('Minimum age in the data set is', df['age'].min()*10, 'and the maximum age is', df['age'].max()*10)

Minimum age in the data set is 66.0 and the maximum age is 109.0


In [28]:
df.to_json('NSMES1988.json ',orient='records')

- Use describe function to understand the data and document at least 2 observations
- Perform memory analysis on the data and recommend what non-default data types you would recommend to optimize memory settings for the dataframe.
- Change the data type to more efficient formats and record the memory usage
- Fix the age and income columns to reflect actual values (check the data dictionary)
- With the fixed data, build a subset for ages over 70
- Export the subset as a new CSV file with a name indicating the new changes and store it in the local space for possible use in other assignments.

In [29]:
df.describe()

Unnamed: 0       visits      nvisits      ovisits     novisits  \
count  4406.000000  4406.000000  4406.000000  4406.000000  4406.000000   
mean   2203.500000     5.774399     1.618021     0.750794     0.536087   
std    1272.046972     6.759225     5.317056     3.652759     3.879506   
min       1.000000     0.000000     0.000000     0.000000     0.000000   
25%    1102.250000     1.000000     0.000000     0.000000     0.000000   
50%    2203.500000     4.000000     0.000000     0.000000     0.000000   
75%    3304.750000     8.000000     1.000000     0.000000     0.000000   
max    4406.000000    89.000000   104.000000   141.000000   155.000000   

         emergency     hospital      chronic          age       school  \
count  4406.000000  4406.000000  4406.000000  4406.000000  4406.000000   
mean      0.263504     0.295960     1.541988     7.402406    10.290286   
std       0.703659     0.746398     1.349632     0.633405     3.738736   
min       0.000000     0.000000     0.000000     6.600000     0.000000   
25%       0.000000     0.000000     1.000000     6.900000     8.000000   
50%       0.000000     0.000000     1.000000     7.300000    11.000000   
75%       0.000000     0.000000     2.000000     7.800000    12.000000   
max      12.000000     8.000000     8.000000    10.900000    18.000000   

            income  
count  4406.000000  
mean      2.527132  
std       2.924648  
min      -1.012500  
25%       0.912150  
50%       1.698150  
75%       3.172850  
max      54.835100

In [30]:
df.info(memory_usage='deep')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4406 entries, 0 to 4405
Data columns (total 20 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Unnamed: 0  4406 non-null   int64  
 1   visits      4406 non-null   int64  
 2   nvisits     4406 non-null   int64  
 3   ovisits     4406 non-null   int64  
 4   novisits    4406 non-null   int64  
 5   emergency   4406 non-null   int64  
 6   hospital    4406 non-null   int64  
 7   health      4406 non-null   object 
 8   chronic     4406 non-null   int64  
 9   adl         4406 non-null   object 
 10  region      4406 non-null   object 
 11  age         4406 non-null   float64
 12  afam        4406 non-null   object 
 13  gender      4406 non-null   object 
 14  married     4406 non-null   object 
 15  school      4406 non-null   int64  
 16  income      4406 non-null   float64
 17  employed    4406 non-null   object 
 18  insurance   4406 non-null   object 
 19  medicaid    4406 non-null  

In [31]:
df['age']= df['age']*10

In [39]:
df['income'] = df['income']*10000

In [ ]:
df['income']= df['income']*10**4
df


Unnamed: 0  visits  nvisits  ovisits  novisits  emergency  hospital  \
0              1       5        0        0         0          0         1   
1              2       1        0        2         0          2         0   
2              3      13        0        0         0          3         3   
3              4      16        0        5         0          1         1   
4              5       3        0        0         0          0         0   
...          ...     ...      ...      ...       ...        ...       ...   
4401        4402      11        0        0         0          0         0   
4402        4403      12        0        0         0          0         0   
4403        4404      10        0       20         0          1         1   
4404        4405      16        1        0         0          0         0   
4405        4406       0        0        0         0          0         0   

         health  chronic      adl region   age afam  gender married  school  \
0       average        2   normal  other  69.0  yes    male     yes       6   
1       average        2   normal  other  74.0   no  female     yes      10   
2          poor        4  limited  other  66.0  yes  female      no      10   
3          poor        2  limited  other  76.0   no    male     yes       3   
4       average        2  limited  other  79.0   no  female     yes       6   
...         ...      ...      ...    ...   ...  ...     ...     ...     ...   
4401    average        0   normal  other  84.0   no  female     yes       8   
4402    average        2   normal  other  78.0   no  female      no      11   
4403    average        5   normal  other  73.0   no    male     yes      12   
4404    average        0   normal  other  66.0   no  female     yes      12   
4405  excellent        0   normal  other  71.0   no    male     yes       0   

        income employed insurance medicaid  
0     28810.00      yes       yes       no  
1     27478.00       no       yes       no  
2      6532.00       no        no      yes  
3      6588.00       no       yes       no  
4      6588.00       no       yes       no  
...        ...      ...       ...      ...  
4401  22497.00       no       yes       no  
4402  58132.00       no       yes       no  
4403  38779.16       no       yes       no  
4404  38779.16       no       yes       no  
4405  65968.00      yes        no       no  

[4406 rows x 20 columns]

In [ ]:
df.info(memory_usage='deep')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4406 entries, 0 to 4405
Data columns (total 20 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Unnamed: 0  4406 non-null   int64  
 1   visits      4406 non-null   int64  
 2   nvisits     4406 non-null   int64  
 3   ovisits     4406 non-null   int64  
 4   novisits    4406 non-null   int64  
 5   emergency   4406 non-null   int64  
 6   hospital    4406 non-null   int64  
 7   health      4406 non-null   object 
 8   chronic     4406 non-null   int64  
 9   adl         4406 non-null   object 
 10  region      4406 non-null   object 
 11  age         4406 non-null   float64
 12  afam        4406 non-null   object 
 13  gender      4406 non-null   object 
 14  married     4406 non-null   object 
 15  school      4406 non-null   int64  
 16  income      4406 non-null   float64
 17  employed    4406 non-null   object 
 18  insurance   4406 non-null   object 
 19  medicaid    4406 non-null  

In [ ]:
num_colm = df.select_dtypes(include= ['int']).columns
num_colm

Index(['Unnamed: 0', 'visits', 'nvisits', 'ovisits', 'novisits', 'emergency',
       'hospital', 'chronic', 'school'],
      dtype='object')

In [ ]:
for col in num_colm:
    df[col] = df[col].astype('int8')

df.info(memory_usage='deep')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4406 entries, 0 to 4405
Data columns (total 20 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Unnamed: 0  4406 non-null   int8   
 1   visits      4406 non-null   int8   
 2   nvisits     4406 non-null   int8   
 3   ovisits     4406 non-null   int8   
 4   novisits    4406 non-null   int8   
 5   emergency   4406 non-null   int8   
 6   hospital    4406 non-null   int8   
 7   health      4406 non-null   object 
 8   chronic     4406 non-null   int8   
 9   adl         4406 non-null   object 
 10  region      4406 non-null   object 
 11  age         4406 non-null   float64
 12  afam        4406 non-null   object 
 13  gender      4406 non-null   object 
 14  married     4406 non-null   object 
 15  school      4406 non-null   int8   
 16  income      4406 non-null   float64
 17  employed    4406 non-null   object 
 18  insurance   4406 non-null   object 
 19  medicaid    4406 non-null  

In [ ]:
mask = df['age']>70
df_over_70 = df[mask]
df_over_70


Unnamed: 0  visits  nvisits  ovisits  novisits  emergency  hospital  \
1              2       1        0        2         0          2         0   
3              4      16        0        5         0          1         1   
4              5       3        0        0         0          0         0   
6              7       9        0        0         0          0         0   
7              8       3        0        0         0          0         0   
...          ...     ...      ...      ...       ...        ...       ...   
4400          49      12        4        1         0          0         0   
4401          50      11        0        0         0          0         0   
4402          51      12        0        0         0          0         0   
4403          52      10        0       20         0          1         1   
4405          54       0        0        0         0          0         0   

         health  chronic      adl   region   age afam  gender married  school  \
1       average        2   normal    other  74.0   no  female     yes      10   
3          poor        2  limited    other  76.0   no    male     yes       3   
4       average        2  limited    other  79.0   no  female     yes       6   
6       average        0   normal  midwest  75.0   no  female      no       8   
7       average        0   normal  midwest  87.0   no  female      no       8   
...         ...      ...      ...      ...   ...  ...     ...     ...     ...   
4400    average        3   normal    other  82.0   no    male     yes       7   
4401    average        0   normal    other  84.0   no  female     yes       8   
4402    average        2   normal    other  78.0   no  female      no      11   
4403    average        5   normal    other  73.0   no    male     yes      12   
4405  excellent        0   normal    other  71.0   no    male     yes       0   

        income employed insurance medicaid  
1     27478.00       no       yes       no  
3      6588.00       no       yes       no  
4      6588.00       no       yes       no  
6      8280.00       no       yes       no  
7     30456.00       no       yes       no  
...        ...      ...       ...      ...  
4400  22497.00       no       yes       no  
4401  22497.00       no       yes       no  
4402  58132.00       no       yes       no  
4403  38779.16       no       yes       no  
4405  65968.00      yes        no       no  

[2838 rows x 20 columns]

In [ ]:
df.to_csv('NSMES1988_NewAgeOver70WithIncome.csv ')

In [ ]:
df.describe()

Unnamed: 0       visits      nvisits      ovisits     novisits  \
count  4406.000000  4406.000000  4406.000000  4406.000000  4406.000000   
mean     -0.156832     5.774399     1.618021     0.692692     0.477985   
std      73.539190     6.759225     5.317056     3.451888     3.459950   
min    -128.000000     0.000000     0.000000  -115.000000  -101.000000   
25%     -64.000000     1.000000     0.000000     0.000000     0.000000   
50%       1.000000     4.000000     0.000000     0.000000     0.000000   
75%      63.000000     8.000000     1.000000     0.000000     0.000000   
max     127.000000    89.000000   104.000000    71.000000   113.000000   

         emergency     hospital      chronic          age       school  \
count  4406.000000  4406.000000  4406.000000  4406.000000  4406.000000   
mean      0.263504     0.295960     1.541988    74.024058    10.290286   
std       0.703659     0.746398     1.349632     6.334050     3.738736   
min       0.000000     0.000000     0.000000    66.000000     0.000000   
25%       0.000000     0.000000     1.000000    69.000000     8.000000   
50%       0.000000     0.000000     1.000000    73.000000    11.000000   
75%       0.000000     0.000000     2.000000    78.000000    12.000000   
max      12.000000     8.000000     8.000000   109.000000    18.000000   

              income  
count    4406.000000  
mean    25271.320468  
std     29246.476178  
min    -10125.000000  
25%      9121.500000  
50%     16981.500000  
75%     31728.500000  
max    548351.000000

In [ ]:
import pandas as pd

In [ ]:
df = pd.read_csv('NSMES1988.csv')

In [35]:
list_cat=df.select_dtypes(include=['object']).columns
print(list_cat)


Index(['health', 'adl', 'region', 'afam', 'gender', 'married', 'employed',
       'insurance', 'medicaid'],
      dtype='object')


In [34]:
list_num=df.select_dtypes(include=['int64','float']).columns
print(list_num)

Index(['Unnamed: 0', 'visits', 'nvisits', 'ovisits', 'novisits', 'emergency',
       'hospital', 'chronic', 'age', 'school', 'income'],
      dtype='object')


In [33]:
# df['age']= df['age']*10
# df['income']= df['income']*10**4
df.groupby('health').agg({ 'age': 'mean', 'income': 'mean', 'visits': 'sum'})

age    income  visits
health                                
average    73.849245  2.551990   19337
excellent  73.102041  3.277411    1176
poor       75.702166  1.905162    4929

- excellent health has the least number of visits
- poor health has the lowest income

In [36]:
pd.crosstab(df['region'], df['health'])

health     average  excellent  poor
region                             
midwest        957         90   110
northeast      694         57    86
other         1237        105   272
west           621         91    86

In [40]:
df.pivot_table(index='region', columns='health', values='income', aggfunc=['mean', 'max'])

mean                                   max            \
health          average     excellent          poor   average excellent   
region                                                                    
midwest    24984.227064  31053.880111  21618.121545  548351.0  113992.0   
northeast  27015.354928  33400.122807  20659.302326  417596.0  186621.0   
other      22010.956023  30024.578190  16851.785184  360024.0  136982.0   
west       31663.772013  37255.872308  21118.850000  242161.0  230179.0   

                     
health         poor  
region               
midwest    115897.0  
northeast  107230.0  
other      146477.0  
west        82140.0

In [ ]:
df.groupby('region').agg({ 'age': 'mean', 'income': 'mean', 'visits': 'sum', 'chronic':'max'})

age        income  visits  chronic
region                                             
midwest    74.076923  25136.342161    6258        8
northeast  73.934289  26797.088793    5099        7
other      73.986989  21662.836976    9003        8
west       74.116541  31165.047494    5082        7

In [ ]:
df['health'].value_counts(normalize=True)

health
average      0.796414
poor         0.125738
excellent    0.077848
Name: proportion, dtype: float64

In [ ]:
df['region'].value_counts(normalize=True)

region
other        0.366319
midwest      0.262596
northeast    0.189968
west         0.181117
Name: proportion, dtype: float64

- most clients are in the "other" region
- midwest has more visits than known regions
- highest income is in the west region

